### DISCUSSION TITLE
How to ad-hoc a shape into the Footer zone?

### DISCUSSION THREAD

**Discussion by MikeAlhayek** [OrchardCMS/OrchardCore#9448] [Q&A] [answered]
I am trying to ad-hoc a shape into the footer zone using a filter. 

I created the following filter and registered it in the IoC.

```
    public class RegisterAsDashboardResourceInjectionFilter : IAsyncResultFilter
    {
        private readonly AdminOptions _adminOptions;
        private readonly IResourceManager _resourceManager;
        private readonly ILayoutAccessor _layoutAccessor;
        private readonly IShapeFactory _shapeFactory;

        public RegisterAsDashboardResourceInjectionFilter(IOptions<AdminOptions> adminOptions,
            IResourceManager resourceManager,
            ILayoutAccessor layoutAccessor,
            IShapeFactory shapeFactory)
        {
            _resourceManager = resourceManager;
            _adminOptions = adminOptions.Value;
            _layoutAccessor = layoutAccessor;
            _shapeFactory = shapeFactory;
        }

        public async Task OnResultExecutionAsync(ResultExecutingContext context, ResultExecutionDelegate next)
        {
            if (context.HttpContext.Request.Path.Equals($"/{_adminOptions.AdminUrlPrefix}", System.StringComparison.OrdinalIgnoreCase) || context.HttpContext.Request.Path.Equals($"/{_adminOptions.AdminUrlPrefix}/", System.StringComparison.OrdinalIgnoreCase))
            {
                await InjectRegisterShapesAsync(_layoutAccessor, _resourceManager, _shapeFactory);
            }

            await next();
        }

        public static async Task InjectRegisterShapesAsync(ILayoutAccessor layoutAccessor, IResourceManager resourceManager, IShapeFactory shapeFactory)
        {
            resourceManager.RegisterResource("stylesheet", RegisterResourceManifest.RegisterResourceName).AtHead();
            resourceManager.RegisterResource("script", RegisterResourceManifest.RegisterResourceName).AtFoot();

            dynamic layout = await layoutAccessor.GetLayoutAsync();

            // The dynamic Layout object will contain a Zones dictionary that you can use to access a zone.
            var footerZone = layout.Zones["Footer"];

            // Add new client dialog
            footerZone.Add(await shapeFactory.New.RegisterModals())
                      .Add(await shapeFactory.New.OnDemandSpinner());

        }
    }
```

My code seems to be similar to [this](https://github.com/Lombiq/Orchard-Training-Demo-Module/blob/dev/Filters/ShapeInjectionFilter.cs)

The code gets called as expected. However, the `RegisterModals.cshtml` and `OnDemandSpinner.cshtml` never get called.

How can I fix this issue?

**Comment by deanmarcussen** [comment, marked as answer]
The code you are using is a bit old

See the layers filter
https://github.com/OrchardCMS/OrchardCore/blob/6fff21e73fb48a50e0b6c404d3bea9e9a606b145/src/OrchardCore.Modules/OrchardCore.Layers/Services/LayerFilter.cs#L139

Lose the `dynamic` and use `await zone.AddAsync(...)`

See also further up on how to correctly determine if you are in an admin layout
https://github.com/OrchardCMS/OrchardCore/blob/6fff21e73fb48a50e0b6c404d3bea9e9a606b145/src/OrchardCore.Modules/OrchardCore.Layers/Services/LayerFilter.cs#L66

(checking the path is no guarantee)

**Reply by MikeAlhayek** [reply]
@deanmarcussen that almost worked. using AddAsync(...) allows me to add single shape. But how can I add two shapes?

The following only adds the last shape and ignores the first one. In the below example, the shape `OnDemandSpinner` gets added while the first one `RegisterModals` is ignored. Its like the second call ignores the first call. If I removed the second call, then the `RegisterModals` shape will get added. But I want to inject both.

```
footerZone.AddAsync(await shapeFactory.New.RegisterModals());
footerZone.AddAsync(await shapeFactory.New.OnDemandSpinner());
```

**Reply by jtkech** [reply]
Yes it may happen if it is not a real zone yet but only a zone on demand without any item yet

In that case on the first AddAsync() do the following that return a real zone

    footerZone = footerZone.AddAsync(await shapeFactory.New.RegisterModals());

**Reply by MikeAlhayek** [reply]
@jtkech That worked! 
```
footerZone = await footerZone.AddAsync(await shapeFactory.New.RegisterModals());
await footerZone.AddAsync(await shapeFactory.New.OnDemandSpinner());
```

Thank you!
